ethcore/light/src/types/request/builder.rs

```diff
@@ -130,12 +138,35 @@ impl Requests {
 
 		self.answered += 1;
 
-		// fill as much of the next request as we can.
-		if let Some(ref mut req) = self.requests.get_mut(self.answered) {
+		// fill as much of each remaining request as we can.
+		for req in self.requests.iter_mut().skip(self.answered) {
```

Comment:
We do iterate over all remaining requests upon each supplied response, but as you guessed it will only request outputs that haven't already been filled.

An optimization would be to just fill the next request upon each supplied response and then do a sweep of all unanswered requests after supplying all responses.